python/strings_udf/strings_udf/lowering.py

```diff
@@ -0,0 +1,637 @@
+)
+
+_string_view_lt = cuda.declare_device(
+    "lt",
+    types.boolean(types.CPointer(string_view), types.CPointer(string_view)),
+)
+
+_string_view_startswith = cuda.declare_device(
+    "startswith",
+    types.boolean(types.CPointer(string_view), types.CPointer(string_view)),
+)
+
+_string_view_endswith = cuda.declare_device(
+    "endswith",
+    types.boolean(types.CPointer(string_view), types.CPointer(string_view)),
+)
+
+_string_view_find = cuda.declare_device(
+    "find",
+    size_type(types.CPointer(string_view), types.CPointer(string_view)),
+)
+
+_string_view_rfind = cuda.declare_device(
+    "rfind",
+    size_type(types.CPointer(string_view), types.CPointer(string_view)),
+)
+
+_string_view_isdigit = cuda.declare_device(
+    "pyisdigit", types.boolean(types.CPointer(string_view), types.int64)
+)
+
+
+_string_view_isalnum = cuda.declare_device(
+    "pyisalnum", types.boolean(types.CPointer(string_view), types.int64)
+)
+
+_string_view_isalpha = cuda.declare_device(
+    "pyisalpha", types.boolean(types.CPointer(string_view), types.int64)
+)
+
+_string_view_isdecimal = cuda.declare_device(
+    "pyisdecimal", types.boolean(types.CPointer(string_view), types.int64)
+)
+
+_string_view_isnumeric = cuda.declare_device(
+    "pyisnumeric", types.boolean(types.CPointer(string_view), types.int64)
+)
+
+_string_view_isspace = cuda.declare_device(
+    "pyisspace", types.boolean(types.CPointer(string_view), types.int64)
+)
+
+_string_view_isupper = cuda.declare_device(
+    "pyisupper", types.boolean(types.CPointer(string_view), types.int64)
+)
+
+_string_view_islower = cuda.declare_device(
+    "pyislower", types.boolean(types.CPointer(string_view), types.int64)
+)
+
+_string_view_count = cuda.declare_device(
+    "pycount",
+    size_type(types.CPointer(string_view), types.CPointer(string_view)),
+)
+
+
+# String function implementations
+def call_len_string_view(st):
+    return _string_view_len(st)
+
+
+@cuda_lower(len, string_view)
+def string_view_len_impl(context, builder, sig, args):
+    sv_ptr = builder.alloca(args[0].type)
+    builder.store(args[0], sv_ptr)
+    result = context.compile_internal(
+        builder,
+        call_len_string_view,
+        nb_signature(size_type, types.CPointer(string_view)),
+        (sv_ptr,),
+    )
+
+    return result
+
+
+def call_string_view_contains(st, substr):
+    return _string_view_contains(st, substr)
+
+
+@cuda_lower(operator.contains, string_view, string_view)
+def string_view_contains_impl(context, builder, sig, args):
+    sv_ptr = builder.alloca(args[0].type)
+    substr_ptr = builder.alloca(args[1].type)
+
+    builder.store(args[0], sv_ptr)
+    builder.store(args[1], substr_ptr)
+    result = context.compile_internal(
+        builder,
+        call_string_view_contains,
+        nb_signature(
+            types.boolean,
+            types.CPointer(string_view),
+            types.CPointer(string_view),
+        ),
+        (sv_ptr, substr_ptr),
+    )
+
+    return result
```

Comment:
I was going to suggest something similar - the one catch is that return types can vary too. Maybe that needs to be an argument of the decorator.

Reply:
This was a great idea! And I did encounter exactly what you mentioned @bdice, with a possible solution in https://github.com/rapidsai/cudf/pull/11319/commits/99b7f3b757f884d90bf16893b97ec5109eb77ce6. I got it down to two decorators: One for binary functions and one for identifier functions, with the single overload of `len` being the odd function out since it's the only unary function that doesn't have the extra table lookup.